File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/null_object/lab/ExpressionVisitor.m
% ExpressionVisitor.m

classdef ExpressionVisitor < handle
	methods
		function visitBinaryMinus(theExpressionVisitor, aBinaryMinus)
			if isempty(aBinaryMinus.Operand1)
				error('Operand1 is empty.');
			end
			aBinaryMinus.Operand1.accept(theExpressionVisitor);
			if isempty(aBinaryMinus.Operand2)
				error('Operand2 is empty.');
			end
			aBinaryMinus.Operand2.accept(theExpressionVisitor);
		end

		function visitBinaryPlus(theExpressionVisitor, aBinaryPlus)
			if isempty(aBinaryPlus.Operand1)
				error('Operand1 is empty.');
			end
			aBinaryPlus.Operand1.accept(theExpressionVisitor);
			if isempty(aBinaryPlus.Operand2)
				error('Operand2 is empty.');
			end
			aBinaryPlus.Operand2.accept(theExpressionVisitor);
		end

		function visitConditional(theExpressionVisitor, aConditional)
			if isempty(aConditional.Operand1)
				error('Operand1 is empty.');
			end
			aConditional.Operand1.accept(theExpressionVisitor);
			if isempty(aConditional.Operand2)
				error('Operand2 is empty.');
			end
			aConditional.Operand2.accept(theExpressionVisitor);
			if isempty(aConditional.Operand3)
				error('Operand3 is empty.');
			end
			aConditional.Operand3.accept(theExpressionVisitor);
		end

		function visitDivide(theExpressionVisitor, aDivide)
			if isempty(aDivide.Operand1)
				error('Operand1 is empty.');
			end
			aDivide.Operand1.accept(theExpressionVisitor);
			if isempty(aDivide.Operand2)
				error('Operand2 is empty.');
			end
			aDivide.Operand2.accept(theExpressionVisitor);
		end

		function visitLiteral(theExpressionVisitor, aLiteral)
		end

		function visitMethodInvocation(theExpressionVisitor, aMethodInvocation)
			aIterator = aMethodInvocation.getOperands;
			while ~aIterator.isDone
				aOperand = aIterator.getCurrent;
				if isempty(aOperand)
					error('aOperand is empty.');
				end
				aOperand.accept(theExpressionVisitor);
				aIterator.advance;
			end
		end

		function visitMultiply(theExpressionVisitor, aMultiply)
			if isempty(aMultiply.Operand1)
				error('Operand1 is empty.');
			end
			aMultiply.Operand1.accept(theExpressionVisitor);
			if isempty(aMultiply.Operand2)
				error('Operand2 is empty.');
			end
			aMultiply.Operand2.accept(theExpressionVisitor);
		end

		function visitUnaryMinus(theExpressionVisitor, aUnaryMinus)
			if isempty(aUnaryMinus.Operand1)
				error('Operand1 is empty.');
			end
			aUnaryMinus.Operand1.accept(theExpressionVisitor);
		end

		function visitUnaryPlus(theExpressionVisitor, aUnaryPlus)
			if isempty(aUnaryPlus.Operand1)
				error('Operand1 is empty.');
			end
			aUnaryPlus.Operand1.accept(theExpressionVisitor);
		end

		function visitVariable(theExpressionVisitor, aVariable)
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/null_object/lab/ExpressionVisitor.java
// ExpressionVisitor.java

public abstract class ExpressionVisitor
{
	public void visit(BinaryMinus aBinaryMinus)
	{
		if (aBinaryMinus.getOperand1() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(BinaryMinus aBinaryMinus) -- aBinaryMinus.getOperand1() == null.");
		}
		aBinaryMinus.getOperand1().accept(this);
		if (aBinaryMinus.getOperand2() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(BinaryMinus aBinaryMinus) -- aBinaryMinus.getOperand2() == null.");
		}
		aBinaryMinus.getOperand2().accept(this);
	}

	public void visit(BinaryPlus aBinaryPlus)
	{
		if (aBinaryPlus.getOperand1() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(BinaryPlus aBinaryPlus) -- aBinaryPlus.getOperand1() == null.");
		}
		aBinaryPlus.getOperand1().accept(this);
		if (aBinaryPlus.getOperand2() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(BinaryPlus aBinaryPlus) -- aBinaryPlus.getOperand2() == null.");
		}
		aBinaryPlus.getOperand2().accept(this);
	}

	public void visit(Conditional aConditional)
	{
		if (aConditional.getOperand1() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(Conditional aConditional) -- aConditional.getOperand1() == null.");
		}
		aConditional.getOperand1().accept(this);
		if (aConditional.getOperand2() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(Conditional aConditional) -- aConditional.getOperand2() == null.");
		}
		aConditional.getOperand2().accept(this);
		if (aConditional.getOperand3() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(Conditional aConditional) -- aConditional.getOperand3() == null.");
		}
		aConditional.getOperand3().accept(this);
	}

	public void visit(Divide aDivide)
	{
		if (aDivide.getOperand1() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(Divide aDivide) -- aDivide.getOperand1() == null.");
		}
		aDivide.getOperand1().accept(this);
		if (aDivide.getOperand2() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(Divide aDivide) -- aDivide.getOperand2() == null.");
		}
		aDivide.getOperand2().accept(this);
	}

	public void visit(Literal aLiteral)
	{
	}

	public void visit(MethodInvocation aMethodInvocation)
	{
		DpIterator<Expression> aIterator = aMethodInvocation.getOperands();
		while (!aIterator.isDone())
		{
			Expression aOperand = aIterator.getCurrent();
			if (aOperand == null)
			{
				throw new DpException("void ExpressionVisitor.visit(MethodInvocation aMethodInvocation) -- aOperand == null.");
			}
			aOperand.accept(this);
			aIterator.advance();
		}
	}

	public void visit(Multiply aMultiply)
	{
		if (aMultiply.getOperand1() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(Multiply aMultiply) -- aMultiply.getOperand1() == null.");
		}
		aMultiply.getOperand1().accept(this);
		if (aMultiply.getOperand2() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(Multiply aMultiply) -- aMultiply.getOperand2() == null.");
		}
		aMultiply.getOperand2().accept(this);
	}

	public void visit(UnaryMinus aUnaryMinus)
	{
		if (aUnaryMinus.getOperand1() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(UnaryMinus aUnaryMinus) -- aUnaryMinus.getOperand1() == null.");
		}
		aUnaryMinus.getOperand1().accept(this);
	}

	public void visit(UnaryPlus aUnaryPlus)
	{
		if (aUnaryPlus.getOperand1() == null)
		{
			throw new DpException("void ExpressionVisitor.visit(UnaryPlus aUnaryPlus) -- aUnaryPlus.getOperand1() == null.");
		}
		aUnaryPlus.getOperand1().accept(this);
	}

	public void visit(Variable aVariable)
	{
	}
}
